### PR TITLE
Fast initial sync

### LINKED ISSUE #479: Improve initial sync
- if no peers, subscribe to peer:connected event and start from there
- pick target `head_slot` + `n` (make n option) slots
- split that range by number of peers and fetch each part from different peer
- once all blocks are obtained, verify only block headers
   - if headers are fine, run state transition with those blocks without any signature verification
   - if it doesn't match fetch blocks again
- repeat until all steps until finalized root encountered

### PR DESCRIPTION
**Pull request by mpetrunic**
resolves #479

### CODE REVIEW

File: packages/lodestar/src/sync/initial.ts
Comment (wemeetagain) [inline packages/lodestar/src/sync/initial.ts]: does this request chunks in parallel? Or is it blocking on each request to finish?

File: packages/lodestar/src/sync/initial.ts
Comment (wemeetagain) [inline packages/lodestar/src/sync/initial.ts]: maybe call this batchRequestBlocks, or something along those lines, we aren't really syncing here, just fetching blocks.

File: packages/lodestar/src/sync/utils/sync.ts
Comment (wemeetagain) [inline packages/lodestar/src/sync/utils/sync.ts]: Maybe call this BlockRange?

**Review by wemeetagain — COMMENTED** [review]
I feel like we should be able to start processing blocks as we receive them, rather than saving them all up until the end.
Also I wonder what the tradeoffs with using the `step` param in beacon_block to interleave requests look like, rather than requesting contiguous ranges.

File: packages/lodestar/src/sync/initial.ts
Comment (mpetrunic) [reply packages/lodestar/src/sync/initial.ts]: in parallel, map will return array of promises to resolve

File: packages/lodestar/src/sync/initial.ts
Comment (mpetrunic) [reply packages/lodestar/src/sync/initial.ts]: yeah already renamed it

File: packages/lodestar/src/sync/utils/sync.ts
Comment (mpetrunic) [reply packages/lodestar/src/sync/utils/sync.ts]: > I feel like we should be able to start processing blocks as we receive them, rather than saving them all up until the end.

If you processing them all at once, you have to verify signatures and everything(as some peers might be malicious and sending wrong blocks). Here we collect them and verify if all block roots match (block.parentRoot) from different peers and execute fast state transition without signature and different root verifications (faster state transition). Target is still small (1 epoch) so we will process blocks in batches of 64 or something

> Also I wonder what the tradeoffs with using the `step` param in beacon_block to interleave requests look like, rather than requesting contiguous ranges.

Not sure, I would leave experimentation with this until we have some stable testnet against which we can benchmark syncing

**Comment by mpetrunic** [comment]
@wemeetagain 
> I feel like we should be able to start processing blocks as we receive them, rather than saving them all up until the end.

If you processing them immediately after receiving, you have to verify signatures and everything(as some peers might be malicious and sending wrong blocks). Here we collect them and verify if all block roots match (block.parentRoot) from different peers and execute fast state transition without signature and different root verifications (faster state transition). Target is still small (1 epoch) so we will process blocks in batches of 64 or something

> Also I wonder what the tradeoffs with using the `step` param in beacon_block to interleave requests look like, rather than requesting contiguous ranges.

Not sure, I would leave experimentation with this until we have some stable testnet against which we can benchmark syncing

**Comment by GregTheGreek** [comment]
Crazy idea what if we do a pre filter, where we do everything but verify the signature... And then process them later

**Comment by mpetrunic** [comment]
> Crazy idea what if we do a pre filter, where we do everything but verify the signature... And then process them later

Could you clarify a bit better

**Comment by wemeetagain** [comment]
> Target is still small (1 epoch) so we will process blocks in batches of 64 or something

I don't see where we're doing that?
It looks like we await all blocks before we start processing blocks. Trying to avoid an OOM error keeping all blocks in memory before processing.

File: packages/lodestar/src/sync/utils/sync.ts
Comment (GregTheGreek) [inline packages/lodestar/src/sync/utils/sync.ts:21]: How recent are the hello's?

**Comment by mpetrunic** [comment]
> > Target is still small (1 epoch) so we will process blocks in batches of 64 or something
> 
> I don't see where we're doing that?
> It looks like we await all blocks before we start processing blocks. Trying to avoid an OOM error keeping all blocks in memory before processing.

https://github.com/ChainSafe/lodestar/pull/509/files#diff-ae79fc115387d8b55109acf5b159e8e2R14

I will add some comments their, so in worst case scenario we are keeping 64 blocks in memory. 
We could optimize this better if we introduce event to signal which block **are we going to** process so we can download next epoch blocks while epoch transition is going on

File: packages/lodestar/src/sync/utils/sync.ts
Comment (mpetrunic) [reply packages/lodestar/src/sync/utils/sync.ts:21]: not recent, I don't think we are refreshing hellos, but I would do this in another PR as it's needed in regular sync as well. Possible solution, every SECONDS_PER_SLOT go trough all peers and refresh hellos.

Comment (GregTheGreek) [reply packages/lodestar/src/sync/utils/sync.ts:21]: Yep!

Intitially i was thinking they would be new because of startup, but the more i thought about it, we could lag at any point need to resync and our hello's our totally out of wack

Another thing to think about is when we actually start syncing. Do we have an ideal `PEER_THRESHOLD` where we hold off on syncing until we have an ideal amount of peers

Comment (mpetrunic) [reply packages/lodestar/src/sync/utils/sync.ts:21]: I put 1 for now, as we will test it against single node most of the time. But I will add that as cli option in this PR so user can configure (I think you can do this in current eth1 nodes).

I left peers as reference to peers in sync module so sync can refresh peer data (and it will reflect in regular and initial sync) without some complicated passing on.

**Comment by mpetrunic** [comment]
It passes unit tests, but we need to improve "trusted" state transition to be able to catch up with other node so this potentially breaks sync. 

We could start dev branch until we merge all those big changes if we care about our current "stable" client?! cc @ChainSafe/eth2-0

File: packages/lodestar/src/chain/chain.ts
Comment (wemeetagain) [inline packages/lodestar/src/chain/chain.ts]: maybe `processedCheckpoint` to match the other events

File: packages/lodestar/src/sync/initial/initial.ts
Comment (wemeetagain) [inline packages/lodestar/src/sync/initial/initial.ts]: We also have a finalizedCheckpoint event. Is that the finalization event we want here?

File: packages/lodestar/src/chain/chain.ts
Comment (mpetrunic) [reply packages/lodestar/src/chain/chain.ts]: sure. 

Side note: we have quite a lot of different event name styles. Do we wan't at some point to have processed:checkpoint or processedCheckpoint

File: packages/lodestar/src/sync/initial/initial.ts
Comment (mpetrunic) [reply packages/lodestar/src/sync/initial/initial.ts]: you would think so. But you will receive finalizedCheckpoint event only after you justify at least finalizedEpoch+1 which will never happen in initial sync

File: packages/lodestar/src/chain/chain.ts
Comment (wemeetagain) [reply packages/lodestar/src/chain/chain.ts]: probably worth making an issue to straighten out the event names.
Maybe `chain:processedCheckpoint` would eventually be the best here.

File: packages/lodestar/src/sync/initial/initial.ts
Comment (ec2) [reply packages/lodestar/src/sync/initial/initial.ts]: That's confusing 😕 ...

File: packages/lodestar/src/sync/initial/initial.ts
Comment (ec2) [inline packages/lodestar/src/sync/initial/initial.ts]: sync = async 😂 Nothing wrong here, just think its funny haha

File: packages/lodestar/src/sync/initial/initial.ts
Comment (mpetrunic) [reply packages/lodestar/src/sync/initial/initial.ts]: think about finalized is really really justified epoch 😄

File: packages/lodestar/src/sync/utils/sync.ts
Comment (wemeetagain) [inline packages/lodestar/src/sync/utils/sync.ts]: I think this should be signingRoot. Maybe call the variable parentRoot

File: packages/lodestar/src/sync/utils/sync.ts
Comment (wemeetagain) [inline packages/lodestar/src/sync/utils/sync.ts]: this should be signingRoot, also we don't need the extra `blockToHeader`, we can signingRoot the BeaconBlock directly.

File: packages/lodestar/src/sync/initial/initial.ts
Comment (wemeetagain) [reply packages/lodestar/src/sync/initial/initial.ts]: can we call this fastSync, or somehow differentiate it from a naive sync?
I think we probably will have support for several initial syncing strategies (eg: fast sync, naive sync, light sync, etc)

File: packages/lodestar/src/sync/utils/sync.ts
Comment (wemeetagain) [inline packages/lodestar/src/sync/utils/sync.ts]: I don't understand the + 1 logic.

I think we need to work with a Checkpoint, ie not throw away the root, because the final part of the fast sync strategy is to check the root against the last downloaded block to ensure we've downloaded the chain we think we've downloaded.

File: packages/lodestar/src/sync/utils/sync.ts
Comment (wemeetagain) [inline packages/lodestar/src/sync/utils/sync.ts]: I think we should change the name here, we use 'target' as a term in FFG votes, makes this kinda confusing.

File: packages/lodestar/src/sync/utils/sync.ts
Comment (wemeetagain) [inline packages/lodestar/src/sync/utils/sync.ts]: I think we should give this name a slightly more descriptive name.
Also, consider adding an end param. (Then we could check the downloaded blocks against a known checkpoint root)
eg:
```typescript
/**
 * Ensure a chain of blocks links properly from start to end
 */
isValidChainOfBlocks(config: IBeaconConfig, start: Hash, end: Hash, blocks: BeaconBlock[]): boolean;
```

File: packages/lodestar/src/sync/utils/sync.ts
Comment (wemeetagain) [inline packages/lodestar/src/sync/utils/sync.ts]: feels sort of strange that this function mutates `blocks` (is relied upon in the calling code)

**Review by wemeetagain — CHANGES_REQUESTED** [review]

File: packages/lodestar/src/sync/utils/sync.ts
Comment (mpetrunic) [reply packages/lodestar/src/sync/utils/sync.ts]: there is no method to fetch non latest finalized checkpoint from peers. I'll extract +1 to constant, basically if you put +2 here it will sync 2 epoch at a time. We can only sync next epoch if there is peer with higher finalized epoch

File: packages/lodestar/src/sync/utils/sync.ts
Comment (mpetrunic) [reply packages/lodestar/src/sync/utils/sync.ts]: again we don't have end hash at that point, only when we reach highest finalized epoch we can check if our checkpoint matches other peers checkpoint

**Comment by mpetrunic** [comment]
@ChainSafe/eth2-0 this is also finished for quite some time

File: packages/lodestar/src/chain/interface.ts
Comment (wemeetagain) [inline packages/lodestar/src/chain/interface.ts]: can we stick with `processedCheckpoint` for consistency for now? Maybe tackle 'robust-ifying' our event names separately.

File: packages/lodestar/src/chain/chain.ts
Comment (wemeetagain) [inline packages/lodestar/src/chain/chain.ts:141]: You can string this through now if you merge master.

File: packages/lodestar/src/sync/initial/fast/index.ts
Comment (wemeetagain) [inline packages/lodestar/src/sync/initial/fast/index.ts:17]: :+1:

**Review by wemeetagain — CHANGES_REQUESTED** [review]

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**
